## Task 1:This task requires to calculate time difference between two timestamps.
### Problem statement :
write a function that takes as input two timestamps of the form 2017/05/13 12:00 and calculates their differences in hours.
Please only return the full hour difference and round the results.

E.g., 2022/02/15 00:05 and 2022/02/15 01:00 would return 1 hour.

- Expanding the above function so that it only calculates the time difference  between 09:00 – 17:00 and only on weekdays.
- Letting the user choose what function to use.
## Solution

We developed fucntions to solve this problem, functions which takes two input values from user in the format "YYYY/MM/DD H:M". first it checks if the format is accordingly, then it promplt to choose 1 or 2. 1 for function to count the hours including all days and 2 for weekdays and only between 09:00 to 17:00.

###Input
User must provide input in this format 2022/02/15 01:00

###Output
It give hours in integers e,g 40.

###Envrionment
Python


In [ ]:
from datetime import datetime, timedelta


def calculate_difference_hourly(time_stamp_1,time_stamp_2):

  """
  This Function takes two parameters and calculate time difference between them
  The Accepted Format : "YYYY/MM/DD H:M"
  It compute hours difference between two provided input and Results a rounded output
  """


  time_stamp_1 = datetime.strptime(time_stamp_1, format) # convert time into string
  time_stamp_2 = datetime.strptime(time_stamp_2, format) # convert time into string
  # compute difference between two timestamps and convert it into hours
  total_hours = round(abs((time_stamp_1 - time_stamp_2).total_seconds() / 3600))
  return total_hours


def calculate_difference_hourly_weekdays(timestamp1, timestamp2):

    """
    This function calculate the time hour difference between two timestamps,
    It also allows to count hours only on weekdays and time from 09:00 to 17:00 (Working hours)

    - checks if user user enters value correctly in accepted format
    - assign lower value to first
    - setting up the specific time hour and weekdays

      Parameters:

      - Timestamp-1 and Timestamp-2
      - return the hours counted (rounded Integer)

    """



    time_1 = datetime.strptime(timestamp1, format)
    time_2 = datetime.strptime(timestamp2, format)

    if time_1 < time_2:
      datetime1 = time_1
      datetime2 = time_2
    else:
      datetime1 = time_2
      datetime2 = time_1

    start_time = datetime.strptime('09:00', '%H:%M').time()
    end_time = datetime.strptime('17:00', '%H:%M').time()

    # Calculate the time difference in hours

    hours = 0
    check_datetime = datetime1

    while check_datetime < datetime2:
        # Removing weekends ,Saturday and Sunday, from counting
        if check_datetime.weekday() < 5:
            # Adjust start and end times for each day
            time_1 = max(datetime.combine(check_datetime.date(), start_time), check_datetime)
            time_2 = min(datetime.combine(check_datetime.date(), end_time), datetime2)
            # Calculate the time difference for the current day in hours
            hour_difference = round(abs((time_1 - time_2).total_seconds() / 3600))
            hours += hour_difference          # Adding the time difference calculated
        check_datetime += timedelta(days=1)   # shifting to the next day

    return hours

if __name__ == '__main__':
  """
  It takes two inputs and checks the format first
  - shows error if the input is not in required format
  - Required another input to use function
  - 1 for function with including  weekends
  - 2 for function only on weekdays and from time 09:00 to 17:00
  """

  print(" --- Hours Calculator --- ")

  # time_stamp_1 = "2023/07/03 08:00"
  # time_stamp_2 = "2023/07/04 08:00"

  format = "%Y/%m/%d %H:%M"
  time_stamp_1 = input("Please enter the 1st timestamp (YYYY/MM/DD HH:MM):") # Taking user input (first)
  time_stamp_2 = input("Please enter the 2nd timestamp (YYYY/MM/DD HH:MM):") # Taking user input (Second)

  excep = False # set to check if program continues

  try:
    datetime.strptime(time_stamp_1, format)# converting into string
    datetime.strptime(time_stamp_2, format)# converting into string
  except ValueError as e:
    print("Invalid timestamp format, Use the format 'YYYY/MM/DD HH:MM' for both timestamps.", e)
    excep = True # If there is error program stops

  if not excep:

    print("1. Select 1 for hours within Weekends Select \n2. Select 2 for weekdays")

    fun = int(input("Select one Value: "))
    if fun == 1: # calling function 1st (including weekends)
      hour_difference = calculate_difference_hourly(time_stamp_1, time_stamp_2)
      print("Total Hours counted for all days : ",hour_difference)

    elif fun==2: # calling function 2nd (without weekends)
      timediff = calculate_difference_hourly_weekdays(time_stamp_1, time_stamp_2)
      print("Total Hours counted for all days : ",timediff)

    else:
      print("Invalid Selection") # if user selects other than 1 and 2 shows error

 --- Hours Calculator --- 
Please enter the 1st timestamp (YYYY/MM/DD HH:MM):2023/07/03 08:00
Please enter the 2nd timestamp (YYYY/MM/DD HH:MM):2023/07/10 17:00
1. Select 1 for hours within Weekends Select 
2. Select 2 for weekdays
Select one Value: 2
Total Hours counted for all days :  48


## Tastk 2: Binary Classification
Task include classification of two classes using a dataset. Two different dataset has been merge on base of 'ID' features which is common in both data.There are in total 19 features for this problem. Many feature has categorical values that needed to convert into numerical values to feed into binary classifier.


## Preprocessing:
-Two different files have been merge using pandas library that was stored into google drive. After analysing data, two features has been dropped, ZIK (for being half with null values) and LIP. Feature NKJUD and LIP are correlated, so to avoid redundancy we need to remove one column.
- To handle categorical features in the dataset, we applied LabelEncoding instead of OneHotEncoding. Since the dataset consists of simple categorical values and does not have complex relationships, LabelEncoding was considered the appropriate choice. By using LabelEncoding, we transformed the categorical values into numerical representations without increasing the dimensionality of the dataset. This approach was efficient in terms of computational cost and suited the characteristics of the dataset.
- To enhance the quality of the dataset, we removed the null values. This preprocessing step was performed to ensure that the data used for analysis and modeling does not contain missing or incomplete information. By removing the null values, we aimed to improve the reliability and accuracy of the subsequent data analysis and modeling processes.
- Dataset has been divided into Train and test split with 20% of test dataset.
- Due to imbalance between label class, we perform oversampling,the purpose of oversampling is to provide the model with sufficient training examples of the minority class, enabling it to learn and generalize better. This approach helps prevent the model from being biased towards the majority class and potentially ignoring important patterns or characteristics of the minority class.

## Model :
SVM is used for this problem because it known for their ability to handle complex decision boundaries by using kernel functions to transform the input space.
Perceptron is also tested but doesnot give high results so we choose SVM for final classification

## Accuracy:
SVM model achieved 89% percent of accuracy, while percetron got 84%.

###Environment

Python, Pandas, Numpy



In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import classification_report, accuracy_score
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter
import os

In [ ]:
path_to_directory= "/content/"
file1 = os.path.join(path_to_directory, "Task1_1.csv")
file2 = os.path.join(path_to_directory, "Task1_2.csv")


In [ ]:
task1 = pd.read_csv(file1, delimiter=';')
task2 = pd.read_csv(file2, delimiter=';')
merge = task1.merge(task2, on=['ID'])
merge

ID  UKL  GJAH  ZIK   HUI   ERZ      CDx BJZHD  NKJUD        LPI  BJKG  \
0        0  160  oooo    x  oooo   www      5.0  vvvv   80.0   800000.0  qqqq   
1        1  153   rrr  NaN   uuu  pppp      0.0   mmm  200.0  2000000.0  qqqq   
2        1  153   rrr  NaN   uuu  pppp      0.0   mmm  200.0  2000000.0  qqqq   
3        2    5  oooo    x  oooo   www     19.0    hh   96.0   960000.0    hh   
4        3    9  oooo  NaN  oooo   www    120.0   kkk    0.0        0.0   qqq   
...    ...  ...   ...  ...   ...   ...      ...   ...    ...        ...   ...   
4470  3695   74  oooo    x  oooo   www  26726.0   sss    NaN        NaN  qqqq   
4471  3696  179  oooo    x  oooo   www      0.0   sss  120.0  1200000.0  qqqq   
4472  3697   82  oooo  NaN  oooo   www    221.0  vvvv    0.0        0.0  qqqq   
4473  3698   13  oooo  NaN  oooo   www      3.0   yyy  100.0  1000000.0  qqqq   
4474  3699   57  oooo    y  oooo   www   4000.0   ggg    0.0        0.0  qqqq   

      POUG    TRE ZUB VOL   UIO VBNM Type    OIN  
0        1  1.750   t   f  uuuu    t    n  17.92  
1        0  0.290   f   f  wwww    f    n  16.92  
2        0  0.290   f   f  wwww    f    n  16.92  
3        1  0.000   f   f  wwww    t    n  31.25  
4        0  0.335   f   f  uuuu    f    n  48.17  
...    ...    ...  ..  ..   ...  ...  ...    ...  
4470     5  2.710   f   t  uuuu    t    y  18.75  
4471     5  8.500   t   t  uuuu    t    y  23.50  
4472    12  4.500   t   t  wwww    t    y  34.17  
4473     5  3.750   t   t  wwww    t    y  27.83  
4474     0  5.000   t   t  uuuu    f    y  36.75  

[4475 rows x 19 columns]

In [ ]:
merge.isna().sum()

ID          0
UKL         0
GJAH       74
ZIK      2608
HUI         0
ERZ        74
CDx         0
BJZHD      77
NKJUD     114
LPI       114
BJKG       77
POUG        0
TRE         0
ZUB         0
VOL         0
UIO        45
VBNM        0
Type        0
OIN        46
dtype: int64

In [ ]:
merge.corr() # checking correlation between each feature

<ipython-input-15-4b7deb99e3cc>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  merge.corr() # checking correlation between each feature


ID       UKL       CDx     NKJUD       LPI      POUG       TRE  \
ID     1.000000  0.012657  0.021023 -0.021661 -0.021661  0.064396  0.068695   
UKL    0.012657  1.000000  0.022318 -0.003121 -0.003121  0.071155  0.054560   
CDx    0.021023  0.022318  1.000000  0.167017  0.167017 -0.011262 -0.006404   
NKJUD -0.021661 -0.003121  0.167017  1.000000  1.000000 -0.147201 -0.216126   
LPI   -0.021661 -0.003121  0.167017  1.000000  1.000000 -0.147201 -0.216126   
POUG   0.064396  0.071155 -0.011262 -0.147201 -0.147201  1.000000  0.289900   
TRE    0.068695  0.054560 -0.006404 -0.216126 -0.216126  0.289900  1.000000   
OIN    0.032020 -0.001687 -0.011533 -0.146436 -0.146436  0.201269  0.514262   

            OIN  
ID     0.032020  
UKL   -0.001687  
CDx   -0.011533  
NKJUD -0.146436  
LPI   -0.146436  
POUG   0.201269  
TRE    0.514262  
OIN    1.000000

In [ ]:
merge.describe(exclude=[np.number]) # exclude numeric values

GJAH   ZIK   HUI   ERZ BJZHD  BJKG   ZUB   VOL   UIO  VBNM  Type
count   4401  1867  4475  4401  4398  4398  4475  4475  4430  4475  4475
unique     3     2     3     3    12     8     2     2     2     2     2
top     oooo     x  oooo   www  vvvv  qqqq     f     t  wwww     t     y
freq    3704  1226  4145  3704   936  2459  2310  3842  2895  2704  4135

In [ ]:
merge['Type'].value_counts() # checking the labels distribution

y    4135
n     340
Name: Type, dtype: int64

In [ ]:
merge = merge.drop(['ZIK','LPI'], axis = 1) # dropping columns ZIK (has many null values) abd LPI, LPI and NKJUB has similar values

In [ ]:
df = merge # making new varaible so if changes are made, it should not in orignal dataset

In [ ]:
from sklearn.preprocessing import OneHotEncoder
columns_to_drop = ['GJAH', 'HUI',	'ERZ', 'BJZHD',	'BJKG',	'ZUB', 'VOL',	'UIO', 'VBNM'] # creating new variable for converting categorical values into onehot Encoding
cat_drop =df.drop(columns= columns_to_drop)
print(df.shape)


data_categorical = df[columns_to_drop]
encoder = OneHotEncoder(sparse_output=False)

encoded_categorical = encoder.fit_transform(data_categorical) # otransforming throug OneHotEncoder
encoded_categorical=np.asarray(encoded_categorical).astype(float) # converting into float values

encoded_categorical_df = pd.DataFrame(encoded_categorical, columns=encoder.get_feature_names_out(columns_to_drop)) # converting into Pandas dataframe

X = pd.concat([cat_drop, encoded_categorical_df], axis=1) # converting into pandas dataframe
X


(4475, 17)


ID  UKL      CDx  NKJUD  POUG    TRE Type    OIN  GJAH_ii  GJAH_oooo  \
0        0  160      5.0   80.0     1  1.750    n  17.92      0.0        1.0   
1        1  153      0.0  200.0     0  0.290    n  16.92      0.0        0.0   
2        1  153      0.0  200.0     0  0.290    n  16.92      0.0        0.0   
3        2    5     19.0   96.0     1  0.000    n  31.25      0.0        1.0   
4        3    9    120.0    0.0     0  0.335    n  48.17      0.0        1.0   
...    ...  ...      ...    ...   ...    ...  ...    ...      ...        ...   
4470  3695   74  26726.0    NaN     5  2.710    y  18.75      0.0        1.0   
4471  3696  179      0.0  120.0     5  8.500    y  23.50      0.0        1.0   
4472  3697   82    221.0    0.0    12  4.500    y  34.17      0.0        1.0   
4473  3698   13      3.0  100.0     5  3.750    y  27.83      0.0        1.0   
4474  3699   57   4000.0    0.0     0  5.000    y  36.75      0.0        1.0   

      ...  BJKG_nan  ZUB_f  ZUB_t  VOL_f  VOL_t  UIO_uuuu  UIO_wwww  UIO_nan  \
0     ...       0.0    0.0    1.0    1.0    0.0       1.0       0.0      0.0   
1     ...       0.0    1.0    0.0    1.0    0.0       0.0       1.0      0.0   
2     ...       0.0    1.0    0.0    1.0    0.0       0.0       1.0      0.0   
3     ...       0.0    1.0    0.0    1.0    0.0       0.0       1.0      0.0   
4     ...       0.0    1.0    0.0    1.0    0.0       1.0       0.0      0.0   
...   ...       ...    ...    ...    ...    ...       ...       ...      ...   
4470  ...       0.0    1.0    0.0    0.0    1.0       1.0       0.0      0.0   
4471  ...       0.0    0.0    1.0    0.0    1.0       1.0       0.0      0.0   
4472  ...       0.0    0.0    1.0    0.0    1.0       0.0       1.0      0.0   
4473  ...       0.0    0.0    1.0    0.0    1.0       0.0       1.0      0.0   
4474  ...       0.0    0.0    1.0    0.0    1.0       1.0       0.0      0.0   

      VBNM_f  VBNM_t  
0        0.0     1.0  
1        1.0     0.0  
2        1.0     0.0  
3        0.0     1.0  
4        1.0     0.0  
...      ...     ...  
4470     0.0     1.0  
4471     0.0     1.0  
4472     0.0     1.0  
4473     0.0     1.0  
4474     1.0     0.0  

[4475 rows x 50 columns]

In [ ]:
columns_to_drop = ['GJAH', 'HUI',	'ERZ', 'BJZHD',	'BJKG',	'ZUB', 'VOL',	'UIO', 'VBNM'] # creating new variable for converting categorical values into onehot Encoding
cat_drop =df.drop(columns= columns_to_drop)
print(df.shape)


data_categorical = df[columns_to_drop]
encoder = le = LabelEncoder()

# Identify categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns

# Apply label encoding to each categorical column
for col in categorical_cols:
    label_encoder = LabelEncoder()
    df[col] = label_encoder.fit_transform(df[col])

X = df

(4475, 17)


In [ ]:
# Creating a instance of label Encoder.
le = LabelEncoder()
X = X.dropna().reset_index(drop=1)
X['Type']= le.fit_transform(X['Type'])

In [ ]:
y = np.array(X['Type']).astype(float)
y = y.reshape(-1,1)
X= X.drop(['Type'], axis=1)
X = np.array(X).astype(float)
X

array([[0.000e+00, 1.600e+02, 1.000e+00, ..., 0.000e+00, 1.000e+00,
        1.792e+01],
       [1.000e+00, 1.530e+02, 2.000e+00, ..., 1.000e+00, 0.000e+00,
        1.692e+01],
       [1.000e+00, 1.530e+02, 2.000e+00, ..., 1.000e+00, 0.000e+00,
        1.692e+01],
       ...,
       [3.697e+03, 8.200e+01, 1.000e+00, ..., 1.000e+00, 1.000e+00,
        3.417e+01],
       [3.698e+03, 1.300e+01, 1.000e+00, ..., 1.000e+00, 1.000e+00,
        2.783e+01],
       [3.699e+03, 5.700e+01, 1.000e+00, ..., 0.000e+00, 0.000e+00,
        3.675e+01]])

In [ ]:
# Checking the shape of input data and label
print(y.shape,y.dtype)
print(X.shape,X.dtype)

(4316, 1) float64
(4316, 16) float64


In [ ]:
y = y.reshape(len(y)) # reshaping the labels for inputting into model
print(y)

[0. 0. 0. ... 1. 1. 1.]


#Splitting, Oversampling and applying classifier


In [ ]:

# Splitting dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# Print the class distribution before oversampling
print("Class distribution before undersampling:", Counter(merge['Type']))
from imblearn.over_sampling import RandomOverSampler

# Create an instance of RandomOverSampler
oversampler = RandomOverSampler(random_state=42)

# Perform undersampling
X_train, y_train = oversampler.fit_resample(X_train, y_train)

# Print the class distribution after undersampling
print("Class distribution after undersampling:", Counter(y_train))

#  SVM classifier
svm_classifier =  SVC()

# Train the SVM classifier
svm_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = svm_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

# Print the results
print("Accuracy:", accuracy)
print("Classification Report:")
print(report)


Class distribution before undersampling: Counter({1: 4135, 0: 340})
Class distribution after undersampling: Counter({0.0: 3185, 1.0: 3185})
Accuracy: 0.8900462962962963
Classification Report:
              precision    recall  f1-score   support

         0.0       0.40      0.98      0.57        63
         1.0       1.00      0.88      0.94       801

    accuracy                           0.89       864
   macro avg       0.70      0.93      0.75       864
weighted avg       0.95      0.89      0.91       864



##Evaluation:
This report provides an evaluation of a SVM classification model's performance based on precision, recall, and F1-score metrics. For class 0.0, the precision is 0.40, indicating that 40% of instances predicted as class 0.0 were correct. The recall is 0.98, meaning that 98% of actual class 0.0 instances were correctly identified. The F1-score is 0.57, representing a harmonic mean of precision and recall. For class 1.0, the precision is 1.00, indicating all instances predicted as class 1.0 were correct. The recall is 0.88, indicating that 88% of actual class 1.0 instances were correctly identified. The F1-score is 0.94. The accuracy of the model is 0.89, meaning 89% of instances were correctly classified. The macro avg precision, recall, and F1-score are 0.70, 0.93, and 0.75, respectively, while the weighted avg precision, recall, and F1-score are 0.95, 0.89, and 0.91, respectively. These metrics provide insights into the model's performance for each class and overall accuracy.